sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ShouldImplementExportedInterfacesBase.cs

```diff
@@ -0,0 +1,126 @@
+                .Any(currentType => currentType.Equals(exportedType));
+
+        protected INamedTypeSymbol GetExportedTypeSymbol(SeparatedSyntaxList<TArgumentSyntax>? attributeArguments,
+            SemanticModel semanticModel)
+        {
+            if (attributeArguments == null)
+            {
+                return null;
+            }
+
+            var arguments = attributeArguments.Value;
+            if (arguments.Count == 0 ||
+                arguments.Count > 2)
+            {
+                return null;
+            }
+
+            var argumentSyntax = GetArgumentFromNamedArgument(arguments) ??
+                GetArgumentFromSingleArgumentAttribute(arguments) ??
+                GetArgumentFromDoubleArgumentAttribute(arguments, semanticModel);
+            var typeOfExpression = GetExpression(argumentSyntax);
+
+            var exportedTypeSyntax = GetExportedTypeSyntax(typeOfExpression);
+            if (exportedTypeSyntax == null)
+            {
+                return null;
+            }
+
+            return semanticModel.GetSymbolInfo(exportedTypeSyntax).Symbol as INamedTypeSymbol;
+        }
+
+        private TArgumentSyntax GetArgumentFromNamedArgument(IEnumerable<TArgumentSyntax> arguments) =>
+            arguments.FirstOrDefault(x => GetIdentifier(x) == "contractType");
+
+        private TArgumentSyntax GetArgumentFromSingleArgumentAttribute(IEnumerable<TArgumentSyntax> arguments)
+        {
+            if (arguments.ElementAtOrDefault(1) != null)
+            {
+                return null;
+            }
+
+            // Only one argument, should be typeof expression
+            return arguments.ElementAtOrDefault(0);
+        }
+
+        private TArgumentSyntax GetArgumentFromDoubleArgumentAttribute(IEnumerable<TArgumentSyntax> arguments,
+            SemanticModel semanticModel)
+        {
+            var firstArgument = GetExpression(arguments.ElementAtOrDefault(0));
+            if (firstArgument != null &&
+                semanticModel.GetConstantValue(firstArgument).Value is string)
+            {
+                // Two arguments, second should be typeof expression
+                return arguments.ElementAtOrDefault(1);
+            }
+
+            return null;
+        }
+
+        protected ITypeSymbol GetAttributeTargetSymbol(SyntaxNode syntaxNode, SemanticModel semanticModel)
+        {
+            // Parent is AttributeListSyntax, we handle only class attributes
+            if (!(syntaxNode.Parent?.Parent is TClassSyntax attributeTarget))
+            {
+                return null;
+            }
+
+            return semanticModel.GetDeclaredSymbol(attributeTarget) as ITypeSymbol;
+        }
+
+        protected abstract string GetIdentifier(TArgumentSyntax argumentSyntax);
+        protected abstract TExpressionSyntax GetExpression(TArgumentSyntax argumentSyntax);
+        protected abstract SyntaxNode GetExportedTypeSyntax(TExpressionSyntax expressionSyntax);
```

Comment:
The purpose of this function is not really obvious... Maybe a small example would help?

Reply:
Adding a comment.